packages/@lwc/engine-dom/src/custom-elements/create-custom-element.ts

```diff
@@ -29,46 +29,55 @@ const nativeLifecycleElementsToUpgradedByLWC = new WeakMap<HTMLElement, boolean>
+        // Do we want to support this? Throw an error? Currently for backwards compat it's a no-op.
+    }
+}
+
+for (const [propName, callback] of entries(LIFECYCLE_CALLBACKS)) {
+    // @ts-expect-error type-mismatch
+    BaseUpgradableConstructor.prototype[propName] = function () {
+        // If the element is in the WeakMap (i.e. it's marked as native lifecycle), and if it was upgraded by LWC,
+        // then it can use native lifecycle
+        if (isTrue(nativeLifecycleElementsToUpgradedByLWC.get(this))) {
+            callback(this);
+        }
+    };
+}
+
 // Creates a constructor that is intended to be used directly as a custom element, except that the upgradeCallback is
 // passed in to the constructor so LWC can reuse the same custom element constructor for multiple components.
 // Another benefit is that only LWC can create components that actually do anything – if you do
 // `customElements.define('x-foo')`, then you don't have access to the upgradeCallback, so it's a dummy custom element.
 // This class should be created once per tag name.
 const createUpgradableConstructor = (isFormAssociated: boolean) => {
-    // TODO [#2972]: this class should expose observedAttributes as necessary
-    class UpgradableConstructor extends HTMLElement {
-        static formAssociated = isFormAssociated;
-
-        constructor(upgradeCallback: LifecycleCallback, useNativeLifecycle: boolean) {
-            super();
-
-            if (useNativeLifecycle) {
-                // When in native lifecycle mode, we need to keep track of instances that were created outside LWC
-                // (i.e. not created by `lwc.createElement()`). If the element uses synthetic lifecycle, then we don't
-                // need to track this.
-                nativeLifecycleElementsToUpgradedByLWC.set(this, elementBeingUpgradedByLWC);
-            }
-
-            // If the element is not created using lwc.createElement(), e.g. `document.createElement('x-foo')`,
-            // then elementBeingUpgradedByLWC will be false
-            if (elementBeingUpgradedByLWC) {
-                upgradeCallback(this);
-            }
-            // TODO [#2970]: LWC elements cannot be upgraded via new Ctor()
-            // Do we want to support this? Throw an error? Currently for backwards compat it's a no-op.
-        }
+    // Note that every class passed into `customElements.define(...)` must be unique, so we can never return
+    // the BaseUpgradableConstructor directly.
+    // https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define#exceptions
+    class UpgradableConstructor extends BaseUpgradableConstructor {}
+    if (isFormAssociated) {
+        // Perf optimization - the vast majority of components have formAssociated=false,
+        // so we can skip the setter in those cases, since undefined works the same as false.
+        // @ts-expect-error type-mismatch
+        UpgradableConstructor.formAssociated = isFormAssociated;
```

Comment:
This optimization (avoiding setting `formAssociated`) is tiny compared to the `extends` optimization, but it still exists. I measured the benefit from this change alone at ~1%:

![Screenshot 2024-03-21 at 3 51 47 PM](https://github.com/salesforce/lwc/assets/283842/573a67d4-d849-4eca-9705-d5f2fbcd853e)